alf/algorithms/tasac_algorithm.py

```diff
@@ -0,0 +1,676 @@
+    "TasacCriticInfo", ["critics", "target_critic", "value_loss"],
+    default_value=())
+
+TasacActorInfo = namedtuple(
+    "TasacActorInfo",
+    ["actor_loss", "action_entropy", "beta_entropy", "adv", "value_loss"],
+    default_value=())
+
+TasacInfo = namedtuple(
+    "TasacInfo",
+    ["action_distribution", "action", "actor", "critic", "alpha", "repeats"],
+    default_value=())
+
+TasacLossInfo = namedtuple('TasacLossInfo', ('actor', 'critic', 'alpha'))
+
+Mode = Enum('AlgorithmMode', ('predict', 'rollout', 'train'))
+
+
+def _safe_categorical(logits, alpha):
+    r"""A numerically stable implementation of categorical distribution
+    :math:`exp(\frac{Q}{\alpha})`.
+    """
+    logits = logits / torch.clamp(alpha, min=1e-10)
+    # logits are equivelant after subtracting a common number
+    logits = logits - torch.max(logits, dim=-1, keepdim=True)[0]
+    return td.Categorical(logits=logits)
+
+
+def _discounted_return(rewards, values, is_lasts, discounts):
+    """Computes discounted return for the first T-1 steps.
+
+    Same with ``tf_agents.utils.value_ops``, this function returns accumulated
+    discounted reward for steps that are StepType.LAST.
+
+    Args:
+        rewards (Tensor): shape is ``[T,B]`` (or ``[T]``) representing rewards.
+        values (Tensor): shape is ``[T,B]`` (or ``[T]``) representing values.
+        is_lasts (Tensor): shape is ``[T,B]`` (or ``[T]``) representing last steps.
+        discounts (Tensor): shape is ``[T,B]`` (or ``[T]``) representing discounts.
+
+    Returns:
+        Tensor: A tensor with shape ``[T-1,B]`` (or ``[T-1]``) representing the
+        discounted returns. Shape is ``[B,T-1]`` when time_major is false.
+    """
+    assert values.shape[0] >= 2, ("The sequence length needs to be "
+                                  "at least 2. Got {s}".format(
+                                      s=values.shape[0]))
+
+    is_lasts = is_lasts.to(dtype=torch.float32)
+    is_lasts = common.expand_dims_as(is_lasts, values)
+    discounts = common.expand_dims_as(discounts, values)
+
+    rets = torch.zeros_like(values)
+    rets[-1] = values[-1]
+    acc_values = rets.clone()
+
+    with torch.no_grad():
+        for t in reversed(range(rewards.shape[0] - 1)):
+            rets[t] = acc_values[t + 1] * discounts[t + 1] + rewards[t + 1]
+            acc_values[t] = is_lasts[t] * values[t] + (
+                1 - is_lasts[t]) * rets[t]
+
+    rets = rets[:-1]
+    return rets.detach()
+
+
+@alf.configurable
+class TASACTDLoss(nn.Module):
+    r"""This TD loss implements the unbiased multi-step Q operator
+    :math:`\mathcal{T}^{\pi^{\text{ta}}}` proposed in the TASAC paper. For a sampled
+    trajectory, it compares the beta action :math:`\tilde{b}_n` sampled from the
+    current policy with the historical rollout beta action :math:`b_n` step by step,
+    and uses the maximum :math:`n` that has :math:`\tilde{b}_n\lor b_n=1` as the
+    target step for boostrapping.
```

Comment:
maximum => minimum?
```
 uses the minimum :math:`n\ge t` that has :math:`\tilde{b}_n\lor b_n=1` as the
    target step for boostrapping the value for step t.
```
Commenting on lines +117 to +118

Reply:
> maximum => minimum?
> 
> ```
>  uses the minimum :math:`n\ge t` that has :math:`\tilde{b}_n\lor b_n=1` as the
>     target step for boostrapping the value for step t.
> ```
> 
> Commenting on lines +117 to +118

Yeah you are right. Should be minimum.